### PR TITLE
Add static-pod-check-workaround utility

### PR DESCRIPTION
**Pull request by donpenney**
The static-pod-check-workaround utility monitors configured static pods. If any containers are found not to be running for a given pod, it is considered a health check failure. If any given pod fails the health check too many consecutive times, kubelet is restarted as a workaround. The checks are only performed once kubelet has been active for a certain period.

### CODE REVIEW

File: static-pod-check/README.md [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+`kubelet.service` if any given pod is deemed unhealthy for too long.
+
+```console
+Usage: static-pod-check-workaround.sh
+Options:
+    --dryrun | -n                 Dry run (don't restart kubelet, log only)
+    --verbose | -v                Turn on debug logs
+    --log-to-stdout               Log to stdout instead of journald
+    --manifests-dir | -m          Location of static pod manifests (default: /etc/kubernetes/manifests)
+    --kubelet-up-min <secs>       Minimum time, in seconds, kubelet must be active
```
Comment (bartwensley) [inline static-pod-check/README.md]: For someone who doesn't use the help, they are going to think this is in minutes not seconds. I'd use --kubelet-up-minimum or --kubelete-up-delay to be safe.

File: static-pod-check/README.md [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+
+If all static pods are healthy, the checks are performed every 300 seconds (configurable with the `--healthy-interval`
+option). If there are unhealthy pods, rechecks are performed every 30 seconds (configurable with the
+`--unhealthy-interval` option).
+
+By default, logs are written to journald with a tag of `static-pod-check-workaround`.
+
+## Installation
+
+The utility can be installed onto a node with an ignition config, writing the script along with a corresponding
```
Comment (bartwensley) [inline static-pod-check/README.md]: I assume this works on master or worker nodes in both SNO and multi-node configurations? May want to say "any master or worker node" or something like that?

File: static-pod-check/README.md [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+  - [Installation](#installation)
+    - [Ignition Config](#ignition-config)
+    - [Site Config ignitionConfigOverride](#site-config-ignitionconfigoverride)
+    - [MachineConfig](#machineconfig)
+  - [Checking Logs](#checking-logs)
+
+## Overview
+
+The `static-pod-check-workaround.sh` utility monitors the state of static pod containers, triggering a restart of
+`kubelet.service` if any given pod is deemed unhealthy for too long.
```
Comment (bartwensley) [inline static-pod-check/README.md:19]: Might be good to point to some publicly available version of the bug this is fixing? Is there a Jira or a PR that could be referenced?

File: static-pod-check/README.md [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+journalctl -u static-pod-check-workaround --no-pager --boot
+
+# Watch logs from static-pod-check-workaround
+journalctl -u static-pod-check-workaround --follow
+```
+
+Logs can be viewed remotely with the following command, with `KUBECONFIG` set appropriately:
+
+```console
+# View logs from static-pod-check-workaround
```
Comment (bartwensley) [inline static-pod-check/README.md]: ... on all master nodes

File: static-pod-check/generate-ignition-config.sh [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+}
+
+function generate_machine_config {
+    local role=$1
+    cat <<EOF
+---
+kind: MachineConfig
+apiVersion: machineconfiguration.openshift.io/v1
+metadata:
+  name: 99-${NAME}-${role}
```
Comment (bartwensley) [inline static-pod-check/generate-ignition-config.sh:55]: We seem to have a bunch of MCs named 99-xxx already. Do they get executed in alphabetical order? I guess it doesn't matter when this one is run?

File: static-pod-check/static-pod-check-workaround.sh [hunk 6]
```diff
@@ -0,0 +1,288 @@
...
+    if ! systemctl -q is-active kubelet.service ; then
+        logdebug "kubelet.service is not active"
+        return 1
+    fi
+
+    local kubelet_start_time
+    local kubelet_start_time_secs
+    local secs_since_start
+
+    kubelet_start_time=$(systemctl show kubelet.service --no-page | awk -F'=' '$1 == "ActiveEnterTimestamp" { print $2}')
```
Comment (bartwensley) [inline static-pod-check/static-pod-check-workaround.sh]: Option is --no-pager

Comment (bartwensley) [inline static-pod-check/static-pod-check-workaround.sh]: This is cleaner:
systemctl show kubelet.service --property ActiveEnterTimestamp --value

File: static-pod-check/README.md [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+  will be restarted. This maximum is configurable with the `--max-failures` option, defaulted to 20 consecutive
+  failures.
+
+If all static pods are healthy, the checks are performed every 300 seconds (configurable with the `--healthy-interval`
+option). If there are unhealthy pods, rechecks are performed every 30 seconds (configurable with the
+`--unhealthy-interval` option).
+
+By default, logs are written to journald with a tag of `static-pod-check-workaround`.
+
+## Installation
```
Comment (bartwensley) [inline static-pod-check/README.md:59]: How can this workaround be removed or disabled? Should that be mentioned in the README as well?

**Review by bartwensley — COMMENTED** [review]
Looks good - just a few minor comments/questions.

File: static-pod-check/README.md (hunk 1 above)
Comment (donpenney) [reply static-pod-check/README.md]: Ok

File: static-pod-check/generate-ignition-config.sh (hunk 5 above)
Comment (donpenney) [reply static-pod-check/generate-ignition-config.sh:55]: I believe so, yes. Order doesn't matter for this one

File: static-pod-check/static-pod-check-workaround.sh (hunk 6 above)
Comment (donpenney) [reply static-pod-check/static-pod-check-workaround.sh]: This does work, though. It allows abbreviated options, as long as the shorthand is unique. "--no-p" would also work here

Comment (donpenney) [reply static-pod-check/static-pod-check-workaround.sh]: perfect!

File: static-pod-check/README.md (hunk 2 above)
Comment (donpenney) [reply static-pod-check/README.md]: ok

File: static-pod-check/README.md (hunk 4 above)
Comment (donpenney) [reply static-pod-check/README.md]: ok

File: static-pod-check/README.md (hunk 7 above)
Comment (donpenney) [reply static-pod-check/README.md:59]: I'll add a section

File: static-pod-check/README.md [hunk 8]
```diff
@@ -187,10 +194,105 @@ spec:
...
+```
+
+> **Note**
+> As disabling the service-unit in this fashion overwrites the systemd service-unit file, re-enabling will require
+> recreating the file by deleting the disable MachineConfig and reinstalling via the method described by
+> [MachineConfig](#machineconfig)
+
+### Installed via MachineConfig: Disabling
+
+To disable the serivce-unit created by MachineConfig, patch the MCs to set enabled to false:
```
Comment (bartwensley) [inline static-pod-check/README.md]: spelling: serivce-unit

File: static-pod-check/README.md (hunk 1 above)
Comment (bartwensley) [reply static-pod-check/README.md]: Looks good

File: static-pod-check/README.md (hunk 2 above)
Comment (bartwensley) [reply static-pod-check/README.md]: Looks good

File: static-pod-check/README.md (hunk 4 above)
Comment (bartwensley) [reply static-pod-check/README.md]: Looks good

File: static-pod-check/generate-ignition-config.sh (hunk 5 above)
Comment (bartwensley) [reply static-pod-check/generate-ignition-config.sh:55]: OK

File: static-pod-check/static-pod-check-workaround.sh (hunk 6 above)
Comment (bartwensley) [reply static-pod-check/static-pod-check-workaround.sh]: Looks good

File: static-pod-check/README.md (hunk 7 above)
Comment (bartwensley) [reply static-pod-check/README.md:59]: Looks good

**Pull request merged**